File: app/src/main/java/com/github/pwittchen/reactivesensors/SensorActivity.java
package com.github.pwittchen.reactivesensors;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.TextView;
import com.github.pwittchen.reactivesensors.library.ReactiveSensors;
import rx.Subscription;

public abstract class SensorActivity extends AppCompatActivity {
  protected int sensorType;
  protected String sensorName;

  private TextView tvSensor;
  private ReactiveSensors reactiveSensors;
  private Subscription subscription;
  private SensorHelper sensorHelper;

  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_sensor_sample);
    tvSensor = (TextView) findViewById(R.id.sensor);
    reactiveSensors = new ReactiveSensors(this);
    sensorHelper = new SensorHelper(reactiveSensors, sensorType, sensorName, tvSensor);
  }

  @Override protected void onResume() {
    super.onResume();

    if (!sensorHelper.deviceHasSensor()) {
      return;
    }

    subscription = sensorHelper.createSubscription();
  }

  @Override protected void onPause() {
    super.onPause();
    sensorHelper.safelyUnsubscribe(subscription);
  }
}

File: app-kotlin/src/main/kotlin/com/github/pwittchen/reactivesensors/kotlinapp/SensorActivity.kt
package com.github.pwittchen.reactivesensors.kotlinapp

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.widget.TextView
import com.github.pwittchen.reactivesensors.library.ReactiveSensors
import rx.Subscription

abstract class SensorActivity : AppCompatActivity() {
    private var subscription: Subscription? = null
    private var sensorHelper: SensorHelper? = null
    protected var sensorType: Int = 0
    protected var sensorName: String = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sensor_sample)
        val textView = findViewById(R.id.sensor) as TextView
        sensorHelper = SensorHelper(ReactiveSensors(this), sensorType, sensorName, textView)
        val sensorHelperWithCast = sensorHelper as SensorHelper //TODO: consider improvement to remove this additional val

        if (!sensorHelperWithCast.deviceHasSensor()) {
            return
        }

        subscription = sensorHelperWithCast.createSubscription()
    }

    override fun onPause() {
        super.onPause()
        if (sensorHelper == null) {
            return
        }

        sensorHelper!!.safelyUnsubscribe(subscription)
    }
}
